### PR TITLE
Initial commit for support vscode.workspace.findFiles feature

### PR DESCRIPTION
**Pull request by vparfonov**
Signed-off-by: Vitalii Parfonov <vparfonov@redhat.com>

Find files across all workspace folders in the workspace. 
More details https://code.visualstudio.com/docs/extensionAPI/vscode-api#_workspace 'findFiles'

### CODE REVIEW

File: packages/plugin-ext/src/main/browser/workspace-main.ts [hunk 1]
```diff
@@ -125,4 +130,31 @@ export class WorkspaceMainImpl implements WorkspaceMain {
         });
     }
 
+    $startFileSearch(includePattern: string, excludePatternOrDisregardExcludes?: string | false,
+                     maxResults?: number, token?: CancellationToken): Promise<UriComponents[]> {
+        const uris: UriComponents[] = new Array();
+        let j = 0;
+        const promises: Promise<any>[] = new Array();
+        for (const root of this.roots) {
+            console.log('******** >> ' + root.uri);
```
Comment (benoitf) [inline packages/plugin-ext/src/main/browser/workspace-main.ts]: hello, I think it needs to be removed

File: packages/plugin-ext/src/main/browser/workspace-main.ts [hunk 2]
```diff
@@ -125,4 +130,31 @@ export class WorkspaceMainImpl implements WorkspaceMain {
...
+            promises[j++] = this.fileSearchService.find(includePattern, {rootUri: root.uri}).then(value => {
+                const paths: string[] = new Array();
+                let i = 0;
+                value.forEach(function (item) {
```
Comment (benoitf) [inline packages/plugin-ext/src/main/browser/workspace-main.ts]: arrow functions

File: packages/plugin-ext/src/api/plugin-api.ts [hunk 3]
```diff
@@ -39,6 +39,7 @@ import {
     DefinitionLink,
     DocumentLink
 } from './model';
+import {CancellationToken} from '@theia/plugin';
```
Comment (benoitf) [inline packages/plugin-ext/src/api/plugin-api.ts]: there is an import
```
 import * as theia from '@theia/plugin';
```

so you should use theia.CancellationToken (without adding this import)

File: packages/plugin-ext/src/plugin/plugin-context.ts [hunk 4]
```diff
@@ -77,6 +77,8 @@ import { TerminalServiceExtImpl } from './terminal-ext';
 import { LanguagesExtImpl, score } from './languages';
 import { fromDocumentSelector } from './type-converters';
 import { DialogsExtImpl } from './dialogs';
+import { GlobPattern} from '@theia/plugin';
```
Comment (benoitf) [inline packages/plugin-ext/src/plugin/plugin-context.ts]: there is already an import
```
import * as theia from '@theia/plugin';
```

File: packages/plugin-ext/src/main/browser/workspace-main.ts [hunk 5]
```diff
@@ -19,24 +19,29 @@ import { WorkspaceExt, MAIN_RPC_CONTEXT, WorkspaceMain, WorkspaceFolderPickOptio
 import { RPCProtocol } from '../../api/rpc-protocol';
 import { WorkspaceService } from '@theia/workspace/lib/browser';
 import Uri from 'vscode-uri';
-import { WorkspaceFoldersChangeEvent, WorkspaceFolder } from '@theia/plugin';
+import { UriComponents } from '../../common/uri-components';
+import { WorkspaceFoldersChangeEvent, WorkspaceFolder, CancellationToken } from '@theia/plugin';
```
Comment (benoitf) [inline packages/plugin-ext/src/main/browser/workspace-main.ts]: please use instead
```
import * as theia from '@theia/plugin';
```

File: packages/plugin-ext/src/plugin/workspace.ts [hunk 6]
```diff
@@ -64,6 +65,40 @@ export class WorkspaceExtImpl implements WorkspaceExt {
...
+                includePattern = include;
+            } else {
+                includePattern = include.pattern;
+            }
+        } else {
+            includePattern = '';
+        }
+
+        let excludePatternOrDisregardExcludes: string | false;
+        if (exclude === null) {
```
Comment (benoitf) [inline packages/plugin-ext/src/plugin/workspace.ts]: do we have somtimes null ? AFAIK in theia we mostly have undefined or values

Comment (evidolob) [reply packages/plugin-ext/src/plugin/workspace.ts]: Yes, we should not use `null` at all

File: packages/plugin-ext/src/main/browser/workspace-main.ts [hunk 7]
```diff
@@ -125,4 +130,31 @@ export class WorkspaceMainImpl implements WorkspaceMain {
...
+                    path = root.uri.endsWith('/') ? root.uri + item : root.uri + '/' + item;
+                    paths[i++] = path;
+                });
+                return Promise.resolve(paths);
+            });
+        }
+        return Promise.all(promises).then(value => {
+            let i = 0;
+                value.forEach(function (path) {
+                    // let path: string;
```
Comment (evidolob) [inline packages/plugin-ext/src/main/browser/workspace-main.ts]: This also should be removed

File: packages/plugin-ext/src/plugin/workspace.ts [hunk 8]
```diff
@@ -14,12 +14,13 @@
  * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
  ********************************************************************************/
 
-import { WorkspaceFolder, WorkspaceFoldersChangeEvent, WorkspaceFolderPickOptions, GlobPattern, FileSystemWatcher } from '@theia/plugin';
+import { WorkspaceFolder, WorkspaceFoldersChangeEvent, WorkspaceFolderPickOptions, GlobPattern, FileSystemWatcher, Uri } from '@theia/plugin';
```
Comment (evidolob) [inline packages/plugin-ext/src/plugin/workspace.ts]: You do not need `Uri` from `'@theia/plugin'`, you can use `URI` that you imported below

File: packages/plugin-ext/src/main/browser/workspace-main.ts (hunk 7 above)
Comment (vparfonov) [reply packages/plugin-ext/src/main/browser/workspace-main.ts]: fixed

File: packages/plugin-ext/src/plugin/plugin-context.ts (hunk 4 above)
Comment (vparfonov) [reply packages/plugin-ext/src/plugin/plugin-context.ts]: fixed

File: packages/plugin-ext/src/plugin/workspace.ts (hunk 8 above)
Comment (vparfonov) [reply packages/plugin-ext/src/plugin/workspace.ts]: fixed

File: packages/plugin-ext/src/plugin/workspace.ts (hunk 6 above)
Comment (vparfonov) [reply packages/plugin-ext/src/plugin/workspace.ts]: fixed

**Comment by benoitf** [comment]
@vparfonov FYI you need to use rebase only instead of "merge with master" in your branch, as it's the only way of merging PR on Theia repo

File: packages/plugin-ext/src/main/browser/workspace-main.ts [hunk 9]
```diff
@@ -125,4 +130,29 @@ export class WorkspaceMainImpl implements WorkspaceMain {
         });
     }
 
+    $startFileSearch(includePattern: string, excludePatternOrDisregardExcludes?: string | false,
+                     maxResults?: number, token?: theia.CancellationToken): Promise<UriComponents[]> {
+        const uris: UriComponents[] = new Array();
```
Comment (benoitf) [inline packages/plugin-ext/src/main/browser/workspace-main.ts]: I think you can use
```
const uris: UriComponents[] = [];
```

**Pull request merged**